### PR TITLE
Handle CDAP Decimal for DBSink

### PR DESCRIPTION
**Pull request by vjosadika**
JIRA Ticket: https://cdap.atlassian.net/browse/PLUGIN-180

### CODE REVIEW

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 1]
```diff
@@ -285,6 +285,11 @@ private void writeToDB(PreparedStatement stmt, Schema.Field field, int fieldInde
           stmt.setTimestamp(sqlIndex, Timestamp.from(record.getTimestamp(fieldName).toInstant()));
           bytesWritten += Long.BYTES;
           break;
+        case DECIMAL:
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java:288]: `writeToDataOut` this method should also be modified

Comment (CuriousVini) [reply database-plugins/src/main/java/io/cdap/plugin/DBRecord.java:288]: what is the existing behavior when database sink writes to DB? Is this change backwards compatible?

Comment (ardian4) [reply database-plugins/src/main/java/io/cdap/plugin/DBRecord.java:288]: **What is the existing behavior when database sink writes to DB? Is this change backwards compatible?**

- Yes this change is backwards compatible. There are unit tests to confirm this, also we have created pipelines to test this. It all depends on the field type of the input structure record schema. 

**`writeToDataOut` this method should also be modified**

- `wirteToDataOut` parameter `(DataOutput out)` accepts only primitive types. For logical type Decimal the type will be BYTES. Please let us know if there is a way on how to proceed to modify the `writeToDataOut` method?

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 2]
```diff
@@ -285,6 +285,11 @@ private void writeToDB(PreparedStatement stmt, Schema.Field field, int fieldInde
...
+          BigDecimal value = record.getDecimal(fieldName);
+          stmt.setBigDecimal(sqlIndex, value);
+          bytesWritten += value.unscaledValue().bitLength() / Byte.SIZE;
```
Comment (rmstar) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: Should we also add the size of the scale (int)?

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java (hunk 1 above)
Comment (CuriousVini) [reply database-plugins/src/main/java/io/cdap/plugin/DBRecord.java:288]: 1. From the code it looks like existing plugin writes decimal types as bytes. After this change, the db sink will write decimal type as `decimal`. Is it correct? 
https://github.com/cdapio/hydrator-plugins/blob/develop/database-plugins/src/main/java/io/cdap/plugin/DBRecord.java#L323

2. I see

Comment (ardian4) [reply database-plugins/src/main/java/io/cdap/plugin/DBRecord.java:288]: Yes

**Review by CuriousVini — APPROVED** [review]

**Comment by CuriousVini** [comment]
@vjosadika Please provide previous behavior to confirm this is not a backwards incompatible change.

**Comment by ardian4** [comment]
> @vjosadika Please provide previous behavior to confirm this is not a backwards incompatible change.

@CuriousVini The previous behavior of DB Sink:
1. Was throwing Exception when we were trying to insert a Decimal value to a non-nullable field in DB.
```
PostgreSQL - old sink using decimal:
java.io.IOException: org.postgresql.util.PSQLException: No value specified for parameter 4.

MySQL - old sink using decimal:
Caused by: java.sql.SQLException: Field 'decimal' doesn't have a default value
```
2. Was storing null value for Decimal values when the field was nullable in DB.

The above two scenarios were happening because, for logical types, there was no case for DECIMAL type and the value for decimal column in the statement was always null (L294).

https://github.com/cdapio/hydrator-plugins/blob/9073013b7d41aee8fbe57df9d70e10efc6bf4209/database-plugins/src/main/java/io/cdap/plugin/DBRecord.java#L269-L295

**Comment by CuriousVini** [comment]
I see.. In that case it seems like existing code is not handling the decimal data correctly.

**Review by CuriousVini — APPROVED** [review]
lgtm, please squash the commits

**Comment by bajram-adapt** [comment]
squashed

**Review by CuriousVini — APPROVED** [review]

**Pull request merged**